pkg/controller/supportbundlecollection/controller.go

```diff
@@ -0,0 +1,746 @@
+		return err
+	}
+	expiredAt := bundle.ObjectMeta.CreationTimestamp.Add(expiredDuration)
+	now := time.Now()
+
+	// Create a CollectionStarted failure condition on the CRD if time is expired. Return nil to avoid the event
+	// to be re-enqueued.
+	if now.Equal(expiredAt) || now.After(expiredAt) {
+		conditions := []v1alpha1.SupportBundleCollectionCondition{
+			{
+				Type:               v1alpha1.CollectionFailure,
+				Status:             metav1.ConditionTrue,
+				Reason:             string(metav1.StatusReasonExpired),
+				LastTransitionTime: metav1.NewTime(time.Now()),
+			},
+			{
+				Type:               v1alpha1.CollectionCompleted,
+				Status:             metav1.ConditionTrue,
+				LastTransitionTime: metav1.NewTime(time.Now()),
+			},
+		}
+		if err := c.addConditions(bundle.Name, conditions); err != nil {
+			klog.ErrorS(err, "Failed to create a CollectionFailure condition")
+			return err
+		}
+		klog.InfoS("SupportBundleCollection is expired", "expiredAt", expiredAt, "now", now)
+		return nil
+	}
+	// Get expected Kubernetes Nodes defined in the CRD.
+	nodeNames, err := c.getBundleNodes(bundle.Spec.Nodes)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get Nodes defined in the support bundle", "name", bundle.Name)
+		return err
+	}
+	// Get expected external Nodes defined in the CRD.
+	externalNodeNames, err := c.getBundleExternalNodes(bundle.Spec.ExternalNodes)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get ExternalNodes defined in the support bundle", "name", bundle.Name)
+		return err
+	}
+	nodeSpan := nodeNames.Union(externalNodeNames)
+	obj, oldInternalBundleExists, _ := c.supportBundleCollectionStore.Get(bundle.Name)
+	if !oldInternalBundleExists {
+		// Get authentication from the Secret provided in authentication field in the CRD
+		authentication, err := c.parseBundleAuth(bundle.Spec.Authentication)
+		if err != nil {
+			klog.ErrorS(err, "Failed to get authentication defined in the SupportBundleCollection CR", "name", bundle.Name, "authentication", bundle.Spec.Authentication)
+			return err
+		}
+		if err := c.addInternalSupportBundleCollection(bundle, nodeSpan, *authentication, metav1.NewTime(expiredAt)); err != nil {
+			klog.ErrorS(err, "Failed to add internal SupportBundleCollection", "name", bundle.Namespace)
+			return err
+		}
+		// Process the support bundle collection when time is up, this will create a CollectionFailure condition if the
+		// bundle collection is not completed in time because any Agent fails to upload the files and does not report
+		// the failure.
+		c.queue.AddAfter(bundle.Name, expiredAt.Sub(now))
+		klog.InfoS("Created internal SupportBundleCollection", "name", bundle.Name)
+	} else {
+		oldInternalBundle := obj.(*types.SupportBundleCollection)
+		if oldInternalBundle.NodeNames.Equal(nodeSpan) {
+			klog.InfoS("The internal SupportBundleCollection span is not changed", "name", bundle.Name)
+			return nil
+		}
```

Comment:
It handles the Node span update case but the controller doesn't watch Node and ExternalNode events, so it can only see NodeSpan change if the collection happens to be reprocessed after a Node event. I think it would cause uncertainty about collection will handle added/updated/deleted node or not.
I feel it should either not handle Node updates at all or watch the two objects. The former is good from supportbundle's perspective IMO. It makes sense to only handle Nodes existing when the collection is handled.